### PR TITLE
Structured data records SalesForce

### PR DESCRIPTION
**Pull request by amcpanaligan**
For #66

### CODE REVIEW

File: src/steps/account/account-create.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import { BaseStep, Field, StepInterface } from '../../core/base-step';
-import { Step, RunStepResponse, FieldDefinition, StepDefinition } from '../../proto/cog_pb';
+import { BaseStep, Field, StepInterface, ExpectedRecord } from '../../core/base-step';
+import { Step, RunStepResponse, FieldDefinition, StepDefinition, RecordDefinition } from '../../proto/cog_pb';
 
 export class CreateAccount extends BaseStep implements StepInterface {
```
Comment (iamEAP) [inline src/steps/account/account-create.ts]: - [ ] Looks like the `AccountAware` mixin didn't get the same field treatment that the other objects got, so the info returned is still quite limited.

Screen Shot 2020-02-16 at 5 16 01 PM

File: src/steps/account/account-create.ts [hunk 2]
```diff
@@ -11,14 +11,25 @@ export class CreateAccount extends BaseStep implements StepInterface {
     type: FieldDefinition.Type.MAP,
     description: 'A map of field names to field values',
   }];
+  protected expectedRecords: ExpectedRecord[] = [{
```
Comment (iamEAP) [inline src/steps/account/account-create.ts]: - [ ] Relatedly, would want to see the same create/modified dates on the account object too:

```
[{
  field: 'CreatedDate',
  type: FieldDefinition.Type.DATETIME,
  description: "Account's Created Date",
}, {
  field: 'LastModifiedDate',
  type: FieldDefinition.Type.DATETIME,
  description: "Account's Last Modified Date",
}],
```

File: src/steps/account/account-delete.ts [hunk 3]
```diff
@@ -15,13 +15,24 @@ export class DeleteAccount extends BaseStep implements StepInterface {
...
+    dynamicFields: false,
+  }];
 
   async executeStep(step: Step): Promise<RunStepResponse> {
     const stepData: any = step.getData().toJavaScript();
 
     try {
       const result = await this.client.deleteAccountByIdentifier(stepData.field, stepData.identifier);
-      return this.pass('Successfully deleted Account with %s %s', [stepData.field, stepData.identifier]);
+      const record = this.keyValue('account', 'Deleted Account', result);
```
Comment (iamEAP) [inline src/steps/account/account-delete.ts]: - [ ] Returned record shouldn't include metadata about the API response that is unrelated to the account itself:

Screen Shot 2020-02-16 at 5 21 17 PM

**Comment by amcpanaligan** [comment]
> Looking good so far, @amcpanaligan! Couple of minor notes inline, though upon looking at the PR in totality, it looks like it's just still WIP (hence no opportunities yet / partial support for accounts so far).

Yes. `hubspot` is more complete. Working on this today

File: src/steps/account/account-field-equals.ts [hunk 4]
```diff
@@ -82,6 +121,24 @@ export class AccountFieldEquals extends BaseStep implements StepInterface {
       return this.error('There was an error during validation of account field: %s', [e.message]);
     }
   }
+
```
Comment (iamEAP) [inline src/steps/account/account-field-equals.ts]: - [ ] Minor note on this section below: recommend limiting use of `titleCase` to just table headers.  Best to keep the key/value record keys (fields) matching the actual field values for possible copy/paste from error message to scenario definition.

File: src/steps/campaign-member/campaign-member-field-equals.ts [hunk 5]
```diff
@@ -64,19 +82,21 @@ export class CampaignMemberFieldEquals extends BaseStep implements StepInterface
       if (!campaignMember) {
         // If no results were found, return a failure.
         return this.fail('No Campaign Membership found between %s and campaign %s', [email, campaignId]);
-      } else if (!campaignMember.hasOwnProperty(field)) {
+      }
+
+      const record = this.keyValue('campaignMember', 'Checked Campaign Member', campaignMember);
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-field-equals.ts]: - [ ] Will want to remove `campaignMember.attributes`

File: src/steps/campaign-member/campaign-member-id-equals.ts [hunk 6]
```diff
@@ -18,16 +18,33 @@ export class CampaignMemberCampaignIdEquals extends BaseStep implements StepInte
...
   async executeStep(step: Step): Promise<RunStepResponse> {
     const stepData: any = step.getData().toJavaScript();
     const email: string = stepData.email;
-    const field: string = 'CampaignId';
     const campaignId: string = stepData.campaignId;
     let campaignMember: Record<string, any>;
 
     try {
-      campaignMember = await this.client.findCampaignMemberByEmailAndCampaignId(email, campaignId, [field]);
+      campaignMember = await this.client.findCampaignMemberByEmailAndCampaignId(email, campaignId);
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-id-equals.ts]: - [ ] Same, re: `campaignMember.attributes`

File: src/steps/contact/contact-field-equals.ts [hunk 7]
```diff
@@ -29,6 +29,24 @@ export class ContactFieldEqualsStep extends BaseStep implements StepInterface {
...
+    }, {
+      field: 'CreatedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Contact's Created Date",
+    }, {
+      field: 'LastModifiedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Contact's Last Modified Date",
+    }],
+    dynamicFields: false,
```
Comment (iamEAP) [inline src/steps/contact/contact-field-equals.ts]: - [ ] Should be `true`.

File: src/steps/contact/contact-field-equals.ts [hunk 8]
```diff
@@ -39,25 +57,24 @@ export class ContactFieldEqualsStep extends BaseStep implements StepInterface {
...
 
     try {
       if (!contact) {
         return this.error('No Contact found with email %s', [email]);
-      } else if (!contact.hasOwnProperty(field)) {
-        return this.error('The %s field does not exist on Contact %s', [field, email]);
-        /* tslint:disable-next-line:triple-equals */
+      }
+
+      const record = this.keyValue('contact', 'Checked Contact', contact);
```
Comment (iamEAP) [inline src/steps/contact/contact-field-equals.ts]: - [ ] Same, re: `contact.attributes`

File: src/steps/campaign-member/campaign-member-field-equals.ts [hunk 9]
```diff
@@ -33,6 +33,24 @@ export class CampaignMemberFieldEquals extends BaseStep implements StepInterface
...
+      description: "Campaign Member's SalesForce ID",
+    }, {
+      field: 'CreatedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Created Date",
+    }, {
+      field: 'LastModifiedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Last Modified Date",
+    }],
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-field-equals.ts]: - [ ] We should also add `CampaignId`, `LeadId`, `ContactId`, and `LeadOrContactId`, since these are all standard to the `CampaignMember` object and they would all be useful as tokens.

File: src/steps/campaign-member/campaign-member-field-equals.ts [hunk 10]
```diff
@@ -33,6 +33,24 @@ export class CampaignMemberFieldEquals extends BaseStep implements StepInterface
...
+    id: 'campaignMember',
+    type: RecordDefinition.Type.KEYVALUE,
+    fields: [{
+      field: 'Id',
+      type: FieldDefinition.Type.STRING,
+      description: "Campaign Member's SalesForce ID",
+    }, {
+      field: 'CreatedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Created Date",
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-field-equals.ts]: - [ ] `Campaign Member's Created Date`

File: src/steps/campaign-member/campaign-member-field-equals.ts [hunk 11]
```diff
@@ -33,6 +33,24 @@ export class CampaignMemberFieldEquals extends BaseStep implements StepInterface
...
+    }, {
+      field: 'LastModifiedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Last Modified Date",
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-field-equals.ts]: - [ ] `Campaign Member's Modified Date`

File: src/steps/campaign-member/campaign-member-id-equals.ts [hunk 12]
```diff
@@ -18,16 +18,33 @@ export class CampaignMemberCampaignIdEquals extends BaseStep implements StepInte
...
+      type: FieldDefinition.Type.STRING,
+      description: "Campaign Member's SalesForce ID",
+    }, {
+      field: 'CreatedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Created Date",
+    }, {
+      field: 'LastModifiedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Last Modified Date",
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-id-equals.ts]: - [ ] Same, re: adding `CampaignId`, `LeadId`, `ContactId`, and `LeadOrContactId`, since these are all standard to the `CampaignMember` object and they would all be useful as tokens.

File: src/steps/campaign-member/campaign-member-id-equals.ts [hunk 13]
```diff
@@ -18,16 +18,33 @@ export class CampaignMemberCampaignIdEquals extends BaseStep implements StepInte
...
+    id: 'campaignMember',
+    type: RecordDefinition.Type.KEYVALUE,
+    fields: [{
+      field: 'Id',
+      type: FieldDefinition.Type.STRING,
+      description: "Campaign Member's SalesForce ID",
+    }, {
+      field: 'CreatedDate',
+      type: FieldDefinition.Type.DATETIME,
+      description: "Lead's Created Date",
```
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-id-equals.ts]: - [ ] `Campaign Member's Created Date`

File: src/steps/campaign-member/campaign-member-id-equals.ts (hunk 12 above)
Comment (iamEAP) [inline src/steps/campaign-member/campaign-member-id-equals.ts]: - [ ] `Campaign Member's Modified Date`

File: src/steps/lead/lead-field-equals.ts [hunk 14]
```diff
@@ -49,19 +67,20 @@ export class LeadFieldEquals extends BaseStep implements StepInterface {
       if (!lead) {
         // If no results were found, return an error.
         return this.error('No Lead found with email %s', [email]);
-      } else if (!lead.hasOwnProperty(field)) {
+      }
+
+      const record = this.keyValue('lead', 'Checked Lead', lead);
```
Comment (iamEAP) [inline src/steps/lead/lead-field-equals.ts]: - [ ] Same, re: `lead.attributes`

File: src/steps/opportunity/opportunity-delete.ts [hunk 15]
```diff
@@ -15,13 +15,24 @@ export class DeleteOpportunity extends BaseStep implements StepInterface {
...
+    dynamicFields: true,
+  }];
 
   async executeStep(step: Step): Promise<RunStepResponse> {
     const stepData: any = step.getData().toJavaScript();
 
     try {
       const result = await this.client.deleteOpportunityByIdentifier(stepData.field, stepData.identifier);
-      return this.pass('Successfully deleted Opportunity with %s %s', [stepData.field, stepData.identifier]);
+      const record = this.keyValue('opportunity', 'Deleted Opportunity', { Id: result.id });
```
Comment (iamEAP) [inline src/steps/opportunity/opportunity-delete.ts]: - [ ] There may be a regression here, because, while the step does appear to actually delete the opportunity, the step appears as though it fails due to possibly passing `undefined` from the Cog to Crank.

Screen Shot 2020-02-25 at 1 50 22 PM

File: src/steps/opportunity/opportunity-field-equals.ts [hunk 16]
```diff
@@ -82,6 +122,24 @@ export class OpportunityFieldEquals extends BaseStep implements StepInterface {
       return this.error('There was an error during validation of opportunity field: %s', [e.message]);
     }
   }
+
+  createRecord(opportunity: Record<string, any>) {
+    Object.keys(opportunity.attributes).forEach(attr => opportunity[titleCase(attr)] = opportunity.attributes[attr]);
```
Comment (iamEAP) [inline src/steps/opportunity/opportunity-field-equals.ts]: - [ ] Same comment as above re: use of `titleCase` on key/value.

File: src/steps/opportunity/opportunity-field-equals.ts [hunk 17]
```diff
@@ -82,6 +122,24 @@ export class OpportunityFieldEquals extends BaseStep implements StepInterface {
...
+    delete opportunity.attributes;
+    return this.keyValue('opportunity', 'Checked Opportunity', opportunity);
```
Comment (iamEAP) [inline src/steps/opportunity/opportunity-field-equals.ts]: - [ ] Strangely, I am still seeing `attributes` come through.

Screen Shot 2020-02-25 at 1 52 50 PM

Comment (amcpanaligan) [reply src/steps/opportunity/opportunity-field-equals.ts]: The `createRecord` was not invoked :(

**Comment by amcpanaligan** [comment]
> Nice! Just a couple more things I noticed in review. And one note for all of the steps:
> 
> * [ ]  All of the records appear to return a `url` field that appears to be metadata from the API response and not actually a field on the object.  We should remove it, similar to the `attributes` property.
> 
> Screen Shot 2020-02-27 at 5 02 05 PM

The `type` and `url` is actually the `attributes` object flatten out. I guess we can remove
